Domestic payments v4

mariantiris opened the pull request:
Issue: https://github.com/SecureApiGateway/SecureApiGateway/issues/1559

Implemented controllers and validators to support v4 domestic payments functionality.
Refactored to account for the package changes in commons and RCS.

dbadham-fr reviewed (commented):
Do we need to have new entities for v4?

I wasn't expecting to see any changes to the Entities / Mongo. As the data gets stored using the FR data model and the model will support both v3 and v4 (contain a superset of the fields from both versions).

mariantiris commented:
You're right @dbadham-fr. I initially thought there will be changes but since we rely on the FR objects and they hold the structure for both v3 and v4, these will not be required anymore. I removed them and switched to using v3 components.

dbadham-fr commented on secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/configuration/validation/DefaultOBValidationModule.java:
```diff
@@ -91,124 +90,182 @@ public class DefaultOBValidationModule {
+    }
+
+   /*@Bean
+    public OBRisk1Validator paymentRiskValidator(@Value("${rs.obie.validation.config.payments
+    .requirePaymentContextCode:false}")
                                                  boolean requirePaymentContextCode) {
         return new OBRisk1Validator(requirePaymentContextCode);
-    }
+    }*/
```
is this needed?

dbadham-fr commented on secure-api-gateway-ob-uk-rs-validation/secure-api-gateway-ob-uk-rs-validation-obie/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/validation/obie/ValidatorsConfiguration.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ComponentScan(basePackageClasses = ValidatorsConfiguration.class)
+public class ValidatorsConfiguration {
+}
```
Can you remove this please?

We want the module: `secure-api-gateway-ob-uk-rs-validation-obie` to be unaware of any spring wiring. The reason for this is that it allows us to customise how validation is done more easily, this is achieved in the class: `DefaultOBValidationModule`.

If we take the component scan approach then it forces everyone to take the validators that spring finds in this package and prevents customers from plugging in custom validation. Currently, if a customer wishes to customise the validation then they can supply their own validation module.

dbadham-fr replied:
Also remove any Spring annotations that have been added to validators

mariantiris replied:
What's your recommended approach on this matter then? Should the beans from DefaultOBValidationModule just be named differently between v3 and v4 or something else? Would that work better?

dbadham-fr replied:
Yes just use different names for the methods, the name of the method will be the bean name which you can use if you need a Qualifier.

dbadham-fr commented on secure-api-gateway-ob-uk-rs-validation/secure-api-gateway-ob-uk-rs-validation-obie/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/validation/obie/v4/payment/BasePaymentRequestValidator.java:
```diff
@@ -0,0 +1,171 @@
+import com.forgerock.sapi.gateway.ob.uk.rs.validation.obie.BaseOBValidator;
+import com.forgerock.sapi.gateway.ob.uk.rs.validation.obie.v4.payment.BasePaymentRequestValidator.PaymentRequestValidationContext;
+
+import uk.org.openbanking.datamodel.v3.error.OBError1;
```
I think this is a reasonable compromise for v4, to continue to use OBError1 from v3 pkg because all exception handling code currently uses this and the schema is the same except changing the code field to be a 4 letter code.

In future we may need to revisit this, in which case it will make sense to introduce our own Error class which can then be converted into v3 and v4 response types in the exception handler. This will be a fairly large piece of work, which is why I think it makes sense to avoid it for now (as there is no functional benefit at the moment).

One thing which needs to be considered is how the OBError1.code value is populated in the response, as in v4 the values are now 4 letter codes e.g. v3 code: `UK.OBIE.Field.Expected` becomes: `U001` - this is a simple mapping according to values here: https://github.com/OpenBankingUK/External_Internal_CodeSets/blob/main/OB_Internal_CodeSet.csv

For ease of migration this translation can be added in the GlobalExceptionHandler - this handler is cross cutting but I expect it has access to the request obj and can therefore determine the API version being used, then it can translate the error code as needed. By convention, the validators can create OBError1 objects with v3 codes, translation is only required for v4 responses.

@mariantiris can you update the v4 functional tests to assert these new error codes please?

FYI - @BohoCode - let me know if you want to discuss this in more detail.

mariantiris commented on secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/exceptions/GlobalExceptionHandler.java:
```diff
@@ -367,4 +373,18 @@ protected ResponseEntity<Object> handleUnexpectedRuntimeException(RuntimeExcepti
+        String path = request.getDescription(false);
+        Pattern pattern = Pattern.compile("/open-banking/v(\\d+)\\.");
+        Matcher matcher = pattern.matcher(path);
+        return matcher.find() && "4".equals(matcher.group(1));
```
Raised this issue for the unit test https://github.com/SecureApiGateway/SecureApiGateway/issues/1572
